### PR TITLE
支持stablelm

### PR DESCRIPTION
**Pull request by learncat163**
# support StableLM 

## 1. 模型概述

StableLM 是 Stability AI 发布的开源语言模型系列。stable-code-3b 是专注于代码生成的 3B 参数模型。

本次 PR 将 StableLM 模型从 HuggingFace transformers 迁移到 PaddleFormers。

## 2. 权重信息

支持直接加载 HuggingFace 权重（从 HF cache 加载）。

也支持paddle版本的权重加载，权重地址：https://aistudio.baidu.com/modelsdetail/47974/

### 3.1 前向对齐概要

使用 float32 精度对齐验证，所有指标满足 1e-2 阈值：

| 指标 | float32 值 | bf16 值 | 阈值 |
|------|-----------|---------|------|
| Top 10 token 一致性 | 10/10 (100%) | 10/10 (100%) | 100% |
| Logits Abs Mean Diff | **0.0000104** | 0.059573 | < 1e-2 |

float32 下 diff < 1e-5，证明代码实现基本正确。bf16 下的 0.05 diff 来自累计误差，但是输出token是一致的

配套有对应的单元测试 `tests/transformers/stablelm/test_modeling.py`的 `test_diff_first_token_logits` 和 `test_diff_generated_tokens` 方法
 
### 3.2 Token 生成一致性

使用原版代码加载原版权重，和Paddle代码加载原版权重，分别进行推理测试，进行输出token的对比。可以看到前30个token的结构都是完全一致的。

3 个 prompt，每一个的前30 个 token 全部一致：

| Prompt | Token 一致匹配数 |
|--------|-------------|
| "猫和狗的区别是什么" | 30/30       |
| "What is the capital of China?" | 30/30       |
| "def fib(n):" | 30/30       |

### 3.3 Logits 精度

受限于模型大小和积累误差，float32 精度下可以实现对齐：

| 方案 | PaddleFormers | 精度 | Logits mean diff | 是否 < 1e-2 |
|------|--------------|------|----------|---------|
| GPU vs GPU | Paddle GPU vs HF GPU | bf16 | 0.059573 | 否 |
| GPU vs GPU | Paddle GPU vs HF GPU | float32 | 0.000006 | 是 |
| CPU vs CPU | Paddle CPU vs HF CPU | float32 | 0.000010 | 是 |

在 float32 模式下均满足精度对齐要求。

## 4. CE验证

目前没有多卡环境，只能运行单卡测试。全参数训练因 OOM (32GB 不足) 未完成，使用 LoRA 训练方式。

### 4.1 SFT LoRA训练

```
per_device_train_batch_size: 1
gradient_accumulation_steps: 1
learning_rate: 0.0001
max_seq_len: 2048
```

| Step | loss | lr | speed |
|------|------|----|-------|
| 1 | 1.6192 | 5.00e-05 | 1.62 steps/s |
| 2 | 1.6722 | 1.00e-04 | 7.80 steps/s |
| 3 | 1.0299 | 8.75e-05 | 10.48 steps/s |
| 4 | 0.9922 | 7.50e-05 | 10.37 steps/s |
| 5 | 0.8924 | 6.25e-05 | 10.36 steps/s |
| 6 | 1.5578 | 5.00e-05 | 10.57 steps/s |
| 7 | 1.2213 | 3.75e-05 | 10.43 steps/s |
| 8 | 1.3143 | 2.50e-05 | 10.48 steps/s |
| 9 | 0.9567 | 1.25e-05 | 10.56 steps/s |
| 10 | 1.0890 | 0.00e+00 | 10.68 steps/s |

### 4.2 热启动测试

从 checkpoint-10 恢复训练到 step 12:

```bash
paddleformers-cli train "${CONFIG_YAML}" \
    model_name_or_path="${MODEL_PATH}" \
    output_dir="${OUTPUT_DIR}" \
    resume_from_checkpoint="${CHECKPOINT_DIR}" \
    max_steps=12 \
    save_steps=1000
```

```
[2026-05-15 19:30:00,408] [INFO] - Starting training from resume_from_checkpoint: checkpoint-10
Step 11: loss=1.0656, lr=1e-05
Step 12: loss=1.0113, lr=0.0
```

可以正常继续之前的训练并输出，loss 连续。

### 4.3 2轮loss对比

| Step | Run 1 Loss | Run 2 Loss | Diff |
|------|------------|------------|------|
| 1 | 1.61921370 | 1.61921370 | 0.0000000000 |
| 2 | 1.67215192 | 1.67215192 | 0.0000000000 |
| 3 | 1.02993941 | 1.02993941 | 0.0000000000 |
| 4 | 0.99218887 | 0.99218887 | 0.0000000000 |
| 5 | 0.89241385 | 0.89241385 | 0.0000000000 |
| 6 | 1.55775344 | 1.55775344 | 0.0000000000 |
| 7 | 1.22126722 | 1.22126722 | 0.0000000000 |
| 8 | 1.31427646 | 1.31427646 | 0.0000000000 |
| 9 | 0.95665663 | 0.95665663 | 0.0000000000 |
| 10 | 1.08901024 | 1.08901024 | 0.0000000000 |

### 4.4 全参微调 OOM

受模型限制，32G的显存不足以全参微调，ms-swift和paddleformers-cli都无法运行。

### 4.5 paddle训练曲线（200步）

| Step | Loss | Step | Loss | Step | Loss |
|------|------|------|------|------|------|
| 1 | 1.6145 | 70 | 0.6405 | 140 | 0.4237 |
| 10 | 1.0364 | 80 | 0.4983 | 150 | 0.3556 |
| 20 | 0.4024 | 90 | 0.4942 | 160 | 0.2598 |
| 30 | 0.5814 | 100 | 0.4305 | 170 | 0.3855 |
| 40 | 1.1273 | 110 | 0.5498 | 180 | 0.6226 |
| 50 | 0.4738 | 120 | 0.4400 | 190 | 0.3783 |
| 60 | 0.5389 | 130 | 0.6519 | 200 | 0.3809 |

### 4.6 PaddleFormers 和 ms-swift LoRA Loss 对比

使用相同配置 (LoRA rank=8, alpha=16, lr=1e-4, batch=1, bf16, max_steps=200, GSM8K) 在两个框架下分别训练，前40步逐步对比（全参对比会OOM）：

| Step | PaddleFormers | ms-swift |
|------|-------------:|---------:|
| 1 | 1.6145 | 1.2934 |
| 2 | 1.7062 | 1.3585 |
| 3 | 1.3543 | 1.1741 |
| 4 | 0.9923 | 1.2285 |
| 5 | 1.3790 | 1.0539 |
| 6 | 1.1833 | 0.9067 |
| 7 | 1.2539 | 0.9102 |
| 8 | 1.4980 | 0.8221 |
| 9 | 0.9900 | 0.7925 |
| 10 | 1.0364 | 0.8365 |
| 11 | 1.0452 | 0.6350 |
| 12 | 1.2762 | 0.6511 |
| 13 | 0.8978 | 0.7907 |
| 14 | 1.0605 | 0.8249 |
| 15 | 0.6361 | 0.7659 |
| 16 | 0.4854 | 0.8246 |
| 17 | 0.9900 | 0.7272 |
| 18 | 0.5569 | 0.6643 |
| 19 | 1.1975 | 0.6725 |
| 20 | 0.4024 | 0.8935 |
| 21 | 0.6554 | 0.5823 |
| 22 | 0.6743 | 0.6764 |
| 23 | 0.3369 | 0.6808 |
| 24 | 0.7735 | 0.6856 |
| 25 | 0.7535 | 0.6679 |
| 26 | 0.6935 | 0.7994 |
| 27 | 0.9044 | 0.7658 |
| 28 | 0.7636 | 0.6856 |
| 29 | 1.0488 | 0.6937 |
| 30 | 0.5814 | 0.7109 |
| 31 | 0.6976 | 0.6762 |
| 32 | 0.8326 | 0.6978 |
| 33 | 0.6736 | 0.7625 |
| 34 | 0.4501 | 0.6830 |
| 35 | 0.5362 | 0.6534 |
| 36 | 0.4856 | 0.6455 |
| 37 | 0.7930 | 0.6472 |
| 38 | 0.6667 | 0.7978 |
| 39 | 0.7641 | 0.6291 |
| 40 | 1.1273 | 0.7705 |

### CODE REVIEW

File: paddleformers/transformers/stablelm/modeling.py [hunk 1]
```diff
@@ -0,0 +1,624 @@
...
+        self.num_key_value_heads = config.num_key_value_heads
+        self.num_key_value_groups = config.num_attention_heads // config.num_key_value_heads
+        self.head_dim = config.hidden_size // config.num_attention_heads
+
+        self.rotary_ndims = int(self.head_dim * config.rope_parameters["partial_rotary_factor"])
+        self.scaling = self.head_dim**-0.5
+        self.attention_dropout = config.attention_dropout
+
+        q_hidden_size = self.num_heads * self.head_dim
+        kv_hidden_size = self.num_key_value_heads * self.head_dim
```
Comment (liuhao2638) [inline paddleformers/transformers/stablelm/modeling.py]: P1 **优先级：P1**

开启 `tensor_model_parallel_size > 1` 时这里会用全局 head 数 reshape 已经被 `ColumnParallelLinear` 切分过的 q/k/v 输出。`GeneralLinear.create(..., tp_plan="colwise")` 在 TP 下默认 `gather_output=False`，每个 rank 只拿到 `out_features / tp`，但当前 `self.num_heads`/`self.num_key_value_heads` 仍是全局值，所以第一层 forward 会在 `reshape([bsz, q_len, self.num_heads, self.head_dim])` 处形状不匹配；同类 Llama/Phi3 实现会先用全局维度建线性层，再把本地 head 数除以 TP degree。请同步这个本地 head 处理，或显式禁用 StableLM 的 TP。

```suggestion
        self.num_heads = config.num_attention_heads
        self.num_key_value_heads = config.num_key_value_heads
        self.num_key_value_groups = config.num_attention_heads // config.num_key_value_heads
        self.head_dim = config.hidden_size // config.num_attention_heads

        q_hidden_size = config.num_attention_heads * self.head_dim
        kv_hidden_size = config.num_key_value_heads * self.head_dim

        if config.tensor_model_parallel_size > 1:
            assert self.num_heads % config.tensor_model_parallel_size == 0, (
                f"num_heads: {self.num_heads}, tensor_model_parallel_size: {config.tensor_model_parallel_size}"
            )
            assert self.num_key_value_heads % config.tensor_model_parallel_size == 0, (
                f"num_key_value_heads: {self.num_key_value_heads}, tensor_model_parallel_size: {config.tensor_model_parallel_size}"
            )
            self.num_heads = self.num_heads // config.tensor_model_parallel_size
            self.num_key_value_heads = self.num_key_value_heads // config.tensor_model_parallel_size

        self.rotary_ndims = int(self.head_dim * config.rope_parameters["partial_rotary_factor"])
        self.scaling = self.head_dim**-0.5
        self.attention_dropout = config.attention_dropout
```

File: paddleformers/transformers/stablelm/modeling.py [hunk 2]
```diff
@@ -0,0 +1,624 @@
...
+            "batch_size": bsz,
+            "seq_length": seq_length,
+            "cache_length": kv_seq_len,
+            "attention_mask": attention_mask,
+            "attn_mask_startend_row_indices": attn_mask_startend_row_indices,
+            "prepare_decoder_attention_mask": self._prepare_decoder_attention_mask,
+        }
+        causal_mask, attn_mask_startend_row_indices = create_causal_mask_and_row_indices(**mask_kwargs)
+
+        hidden_states = inputs_embeds
```
Comment (liuhao2638) [inline paddleformers/transformers/stablelm/modeling.py:539]: P1 **优先级：P1**

`StableLmDecoderLayer` 已把 LayerNorm/Linear 标成 `input_is_parallel=config.sequence_parallel`，但 `StableLmModel.forward` 没有像 Llama/Qwen/Phi3 那样在 embedding 后执行 `ScatterOp.apply`。因此用户配置 `sequence_parallel=True` 时，后续 `ColumnSequenceParallelLinear`/`RowSequenceParallelLinear` 会按已切分的序列输入工作，而这里仍传入完整 `[bsz, seq, hidden]`，多卡训练/推理会得到错误的张量布局或通信形状错误。请在设置 `hidden_states` 前按现有模型模式 scatter，并同步 attention 中 sequence-parallel 的 batch/seq reshape。

可按下面的形状处理补齐（还需要在文件顶部引入 `ScatterOp`，并在 attention 输出后按 `config.sequence_parallel` reshape）：

```python
from paddle.distributed.fleet.utils.sequence_parallel_utils import ScatterOp

...
if self.config.sequence_parallel:
    inputs_embeds = inputs_embeds.reshape([-1, inputs_embeds.shape[-1]])
    inputs_embeds = ScatterOp.apply(inputs_embeds)

hidden_states = inputs_embeds
```

**Review by liuhao2638 — CHANGES_REQUESTED** [review]
已完成代码 review，发现 StableLM 在多卡 tensor parallel / sequence parallel 路径上还有需要合入前修复的问题，具体已放在行级评论中。

当前 `unittest-gpu-ci` 失败停在 `Change python version` 阶段，未进入测试执行，建议 rerun 后再确认完整 CI。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — APPROVED** [review]
已复查新提交。StableLM 的 tensor parallel head 拆分和 sequence parallel scatter/reshape 已按现有模型模式补齐；当前已展示的检查为通过状态，未发现需要阻塞合入的问题。本次没有新增行级评论。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: paddleformers/transformers/stablelm/modeling.py [hunk 3]
```diff
@@ -0,0 +1,645 @@
...
+        seq_len: Optional[int] = None,
+    ) -> tuple["paddle.Tensor", float]:
+        base = config.rope_parameters["rope_theta"]
+        partial_rotary_factor = config.rope_parameters.get("partial_rotary_factor", 1.0)
+        head_dim = getattr(config, "head_dim", None) or config.hidden_size // config.num_attention_heads
+        dim = int(head_dim * partial_rotary_factor)
+
+        attention_factor = 1.0
+
+        inv_freq = 1.0 / (base ** (paddle.arange(0, dim, 2, dtype=paddle.int64).astype(dtype=paddle.float32) / dim))
```
Comment (risemeup1111) [inline paddleformers/transformers/stablelm/modeling.py:77]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

当前 `Lint` 检查失败，CI 日志显示 `black` 会重写 `paddleformers/transformers/stablelm/modeling.py`（`reformatted paddleformers/transformers/stablelm/modeling.py`）。这会阻塞合入；请按仓库的 Black 配置格式化该文件并提交格式化结果。

可直接执行：

```bash
python -m black paddleformers/transformers/stablelm/modeling.py
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查最新提交。此前 StableLM 的 TP/SP 代码问题在当前代码中仍保持已修复状态；这次新增阻塞点是当前 `Lint` 检查失败，`black` 会重写 `paddleformers/transformers/stablelm/modeling.py`。具体处理要求已放在行级评论中，请格式化后提交新的 commit。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — APPROVED** [review]
已复查新的格式化提交。此前 P1 格式化问题已通过新 commit 修复，当前 `Lint` 检查已通过；本轮未发现新的阻塞问题。仍在运行中的 CI 请以最终结果为准。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — APPROVED** [review]
已复查最新提交。此次改动移除了会在普通 CI 中在线拉取 ModelScope tokenizer 的非 slow 用例，之前的 P1 格式化问题仍保持修复；当前未发现新的阻塞问题。仍在运行中的 CI 请以最终结果为准。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>
